### DISCUSSION TITLE
Compiling custom firmware for esp32c3

### DISCUSSION THREAD

**Discussion by skeatz** [micropython/micropython#15648] [ESP32] [answered]
I am trying to add custom modules for the esp32c3 build and compile it to a firmware.bin file. I have tried google and searching the micropython forums but could not find any clear answers to some questions that I have.

-  can I just add the required python modules e.g. tm1637.py, sdcard.py to the ports/esp32/modules and compile the firmware for the c3 or should I add it to modules folder for the specific board that I will be using, e.g. ports/esp32/boards/LOLIN_C3_MINI/modules?
- I noticed that in the ports/esp32/boards/ folder, there is a manifest.py file with the following code:
```py
# Require some micropython-lib modules.
require("aioespnow")
require("dht")
require("ds18x20")
require("neopixel")
require("onewire")
require("umqtt.robust")
require("umqtt.simple")
require("upysh")
```
- can I just add the modules that I need to the requirements here? How do I know what modules are available? I have tried searching the micropython source code and there doesn't seem to be a specific folder where I see what library modules are available.

Apologies if this question has been asked before. I have tried searching and didn't find any similar questions. Any help or pointers would be most appreciated. Thanks

**Comment by ricksorensen** [comment, marked as answer]
see: [https://docs.micropython.org/en/latest/reference/manifest.html](https://docs.micropython.org/en/latest/reference/manifest.html)
The `require` commands will in `micropython-lib` for the modules.  A good practice would be to define your custom mahifest file outside of the release, and include it with:
`make BOARD=MYBOARD FROZEN_MANIFEST=/path/to/my/project/manifest.py`
see `module` command.

**Reply by skeatz** [reply]
@ricksorensen: Thanks. Will take a look at the documentation link and try building my custom firmware over the next couple of days. 🙏
